frontend/ui/lib/widgets/retryable_async_loadable.dart

```diff
@@ -0,0 +1,88 @@
+import 'package:flutter/material.dart';
+
+/// Provides loading and retrying functionality.
+/// This should only be used with an idempotent async function.
+class RetryableAsyncLoadable<T> extends StatefulWidget {
+  final _OnLoad<T> onLoad;
+  final _LoadFunction<T> loadFunction;
+  final String errorMessage;
+
+  RetryableAsyncLoadable({
+    key,
+    @required this.onLoad,
+    @required this.loadFunction,
+    @required this.errorMessage,
+  }) : super(key: key);
+
+  @override
+  State createState() =>
+      _RetryableAsyncLoadableState(onLoad, loadFunction, errorMessage);
+}
+
+typedef _LoadFunction<T> = Future<T> Function();
+typedef _OnLoad<T> = Widget Function(T);
+
+class _RetryableAsyncLoadableState<T> extends State<RetryableAsyncLoadable<T>> {
```

Comment:
I'm very disappointed that Dart types seem to be covariant without respect to safety... 😞

Reply:
What do you mean by that? is there something I'm missing with the type system here?

Reply:
Disclaimer: apparently I don't know what I'm talking about because the [`State` type parameter signature](https://github.com/flutter/flutter/blob/bb3280885c8d19449f2089ab2d98e312430ed28a/packages/flutter/lib/src/widgets/framework.dart#L1036) is apparently `T extends StatefulWidget`, which is presumably responsible for the variance.

I was reading through the Dart developer docs and came across [this](https://dart.dev/faq#q-why-are-generics-covariant). Allowing covariants by default (across the board) is unsafe. (There are situations where covariance/contravariance are both safe and desired, but it's subtle to get correct and generally preferable for the compiler to enforce it appropriately, as done in Java by wildcards and in Scala natively.)

The following program compiles but fails at runtime due to spurious variance:

```dart
void main() {
    Generic<Bar> f = Generic<Bar>();
    Generic<Foo> g = f;
    // The following line fails due to the underlying storage type
    // (Generic<Bar>) expecting the Bar subtype.
    g.set(Foo());
}

class Generic<T> {
    T _value;

    T get() {
        return _value;
    }

    void set(T value) {
        _value = value;
    }
}

class Foo {}

class Bar extends Foo {}
```

You can imagine non-trivial programs in which this can causes data corruption (or, at best, the equivalent of runtime ClassCastExceptions, which is apparently what happens here):

```
Unhandled exception:
type 'Foo' is not a subtype of type 'Bar' of 'value'
#0      Generic.set (file:///private/tmp/hello/main.dart:0)
#1      main (file:///private/tmp/hello/main.dart:6)
#2      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:301)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168)
```

Anyway, the punchline is that I saw that `StatefulWidget` had a method `State<StatefulWidget> createState()` and assumed that this only compiled because it took advantage of sloppy variance. However, it looks like `State` itself is declared in such a way that its variance is explicit (and its use is presumably safe). Doesn't change the fact that unchecked variance is dangerous (incorrect), but it's actually fine here. 😉

Reply:
Ah, sketch. I will say though with how Flutter works I don't think it will ever matter.

The only thing I could see that potentially mattering with is for subtypes with the same key. Flutter identifies widgets in the rendering tree by type and key when they get moved around after the initial rendering. IDK if it uses exact type signatures or covariant type signatures though, and I don't think this application will often have the same key for covariant subtyped `RetryableAsyncLoadable`s.